src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -413,6 +474,70 @@ public static UIntPtr CompareExchange(ref UIntPtr location1, UIntPtr value, UInt
             return (UIntPtr)CompareExchange(ref Unsafe.As<UIntPtr, int>(ref location1), (int)value, (int)comparand);
 #endif
         }
+
+        /// <summary>Compares two instances of the specified type <typeparamref name="T"/> for equality and, if they are equal, replaces the first one.</summary>
+        /// <param name="location1">The destination, whose value is compared with <paramref name="comparand"/> and possibly replaced.</param>
+        /// <param name="value">The value that replaces the destination value if the comparison results in equality.</param>
+        /// <param name="comparand">The object that is compared to the value at <paramref name="location1"/>.</param>
+        /// <returns>The original value in <paramref name="location1"/>.</returns>
+        /// <exception cref="NullReferenceException">The address of <paramref name="location1"/> is a null pointer.</exception>
+        /// <exception cref="NotSupportedException">An unsupported <typeparamref name="T"/> is specified.</exception>
+        /// <typeparam name="T">
+        /// The type to be used for <paramref name="location1"/>, <paramref name="value"/>, and <paramref name="comparand"/>.
+        /// This type must be a reference type, an enum type (i.e. typeof(T).IsEnum is true), or a primitive type (i.e. typeof(T).IsPrimitive is true).
+        /// </typeparam>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [return: NotNullIfNotNull(nameof(location1))]
+        public static T CompareExchange<T>(ref T location1, T value, T comparand)
```

Comment:
This should probably get re-added to the JIT intrinsic.

Reply:
I likely will, but wanted to see the impact on codegen

Reply:
If you want to optimize it for CoreCLR, putting the non `class` path in a separate local method might be slightly better for avoiding regressions with the inliner after you re-add `Intrinsic` since it should then almost never be reached (only for debug and indirect calls I think).